### PR TITLE
[luci] Add INT4 weights quantization to quantizer

### PR DESCRIPTION
**Pull request by SlavikMIPT**
This commit adds INT4 weights quantization to QuantizeWeightsOnly.cpp.

ONE-DCO-1.0-Signed-off-by: Vyacheslav Bazhenov <slavikmipt@gmail.com>

### CODE REVIEW

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 1]
```diff
@@ -100,15 +100,24 @@ void symmetric_wquant_with_minmax_per_layer(CircleConst *node, float min, float
       std::min(kMaxScale, std::max(kMinScale, quantized_values[i]));
   }
 }
+int32_t compute_sym_max_scale(const loco::DataType &type)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizationUtils.cpp:103]:
```suggestion
}

int32_t compute_sym_max_scale(const loco::DataType &type)
```

File: compiler/luci/pass/src/QuantizationUtils.cpp [hunk 2]
```diff
@@ -100,15 +100,24 @@ void symmetric_wquant_with_minmax_per_layer(CircleConst *node, float min, float
...
+{
+  if (type == loco::DataType::S4)
+    return std::numeric_limits<int8_t>::max() >> 4;
+  else if (type == loco::DataType::S8)
+    return std::numeric_limits<int8_t>::max();
+  else
+    return std::numeric_limits<int16_t>::max();
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.cpp:110]:
```suggestion
  else if (type == loco::DataType::S16)
    return std::numeric_limits<int16_t>::max();
  else
    throw std::runtime_error("Unsupported dtype for symmetric quantization")
```

File: compiler/luci/pass/src/QuantizationUtils.h [hunk 3]
```diff
@@ -23,6 +23,9 @@
 namespace luci
 {
 
+// Compute max scale using given data type for symmetric quantization (int4/int8/int16)
+int32_t compute_sym_max_scale(const loco::DataType &type);
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizationUtils.h:27]: Could you rename the function?
```suggestion
// Return the max value of dtype for symmetric quantization (int4/int8/int16)
int32_t max_for_symm_quant(const loco::DataType type);
```

**Review by jinevening — APPROVED** [review]
LGTM

**Review by seanshpark — APPROVED** [review]
LGTM

**Pull request merged**